lib/config/config-cache.js

```diff
@@ -0,0 +1,150 @@
+/**
+ * @fileoverview Responsible for caching config files
+ * @author Sylvan Mably
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Constants
+//------------------------------------------------------------------------------
+
+const VECTOR_SEP = ",";
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+/**
+ * Get a string hash for a vector object
+ * @param {Array<number|string>} vector config vector to hash
+ * @returns {string} hash of the vector values
+ * @private
+ */
+function hash(vector) {
+    return vector ? vector.join(VECTOR_SEP) : "";
+}
+
+//------------------------------------------------------------------------------
+// API
+//------------------------------------------------------------------------------
+
+/**
+ * Configuration caching class (not exported)
+ */
+class ConfigCache {
+
+    constructor() {
+        this.filePathCache = new Map();
+        this.localHierarchyCache = new Map();
+        this.mergedVectorCache = new Map();
+        this.mergedCache = new Map();
+    }
+
+    /**
+     * Empties all the subcaches.
+     * @returns {void}
+     * @private
+     */
+    reset() {
+        this.filePathCache.clear();
+        this.localHierarchyCache.clear();
+        this.mergedVectorCache.clear();
+        this.mergedCache.clear();
+    }
+
+    /**
+     * Gets a config object from the cache for the specified config file path.
+     * @param {string} configFilePath the absolute path to the config file
+     * @returns {Object|null} config object, if found in the cache, otherwise null
+     * @private
+     */
+    getConfig(configFilePath) {
+        return this.filePathCache.get(configFilePath);
+    }
+
+    /**
+     * Sets a config object in the cache for the specified config file path.
+     * @param {string} configFilePath the absolute path to the config file
+     * @param {Object} config the config object to add to the cache
+     * @returns {void}
+     * @private
+     */
+    setConfig(configFilePath, config) {
+        this.filePathCache.set(configFilePath, config);
+    }
+
+    /**
+     * Gets a list of hierarchy-local config objects that apply to the specified directory.
```

Comment:
Configs parsed from config files in the filesystem hierarchy that apply to files in directories below them (as opposed to, say, the user's personal config from their homedir, or a config file passed as on option on the command line). Totally willing to change the terminology if you can think of a better way to describe these.